src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -80,75 +93,78 @@ class ArrayDeque[A] private[ArrayDeque](
+    if (elems.nonEmpty) {
+      ArrayDeque.knownSize(elems) match {
+        case srcLength if srcLength >= 0 && 2*(srcLength + this.length) >= mask =>
+          val finalLength = srcLength + this.length
+          val array2 = ArrayDeque.alloc(finalLength)
+          elems.copyToArray(array2.asInstanceOf[Array[A]])
+          copySliceToArray(srcStart = 0, dest = array2, destStart = srcLength, maxItems = size)
+          resetInternal(array = array2, start = 0, end = finalLength)
+        case _ =>
+          val copy = clone()
+          end = start
+          this ++= elems ++= copy
+      }
     }
     this
   }
 
   override def insertAll(idx: Int, elems: scala.collection.Traversable[A]): Unit = {
     ArrayDeque.checkIndex(idx, this)
-    if (elems.isEmpty) return
-    elems.sizeHintIfCheap match {
-      case srcLength if srcLength >= 0 =>
-        val finalLength = srcLength + this.length
-        // Either we resize right away or move prefix right or suffix left
-        if (2*finalLength >= mask) {
-          val array2 = ArrayDeque.alloc(finalLength)
-          copySliceToArray(srcStart = 0, dest = array2, destStart = 0, maxItems = idx)
-          elems.copyToArray(array2.asInstanceOf[Array[A]], idx)
-          copySliceToArray(srcStart = idx, dest = array2, destStart = idx + srcLength, maxItems = size)
-          set(array = array2, start = 0, end = finalLength)
-        } else {
+    if (elems.nonEmpty) {
+      ArrayDeque.knownSize(elems) match {
+        case srcLength if srcLength >= 0 =>
+          val finalLength = srcLength + this.length
+          // Either we resize right away or move prefix right or suffix left
+          if (2*finalLength >= mask) {
+            val array2 = ArrayDeque.alloc(finalLength)
+            copySliceToArray(srcStart = 0, dest = array2, destStart = 0, maxItems = idx)
+            elems.copyToArray(array2.asInstanceOf[Array[A]], idx)
+            copySliceToArray(srcStart = idx, dest = array2, destStart = idx + srcLength, maxItems = size)
+            resetInternal(array = array2, start = 0, end = finalLength)
+          } else {
+            val suffix = drop(idx)
+            end = start_+(idx)
+            elems.foreach(appendAssumingCapacity)
+            suffix.foreach(appendAssumingCapacity)
+          }
+        case _ => //expensive to compute size
           val suffix = drop(idx)
           end = start_+(idx)
-          elems.foreach(appendAssumingCapacity)
-          suffix.foreach(appendAssumingCapacity)
-        }
-      case _ => //expensive to compute size
-        val suffix = drop(idx)
-        end = start_+(idx)
-        this ++= elems ++= suffix
+          this ++= elems ++= suffix
+      }
     }
   }
 
   override def remove(idx: Int, count: Int): Unit = {
+    if (count < 0) throw new IllegalArgumentException(s"removing negative number of elements: $count")
     ArrayDeque.checkIndex(idx, this)
-    if (count <= 0) return
+    if (count == 0) return
```

Comment:
Checking the index should come after one knows there's something to remove (as in current collections).  Also, the logic should probably be
```
if (count <= 0) {
  if (count < 0) throw ...
  else return
}
```
just in case the JIT compiler doesn't rewrite it that way for efficiency on the success/do-something case.